mirgecom/float_comparisons.py

```diff
@@ -0,0 +1,70 @@
+""":mod:`mirgecom.float_comparisons` provides comparisons for float-valued arrays.
+
+Comparison Functions
+^^^^^^^^^^^^^^^^^^^^
+.. autofunction:: componentwise_norm
+.. autofunction:: componentwise_err
+.. autofunction:: within_tol
+"""
+
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+from mirgecom.fluid import ConservedVars
+from pytools.obj_array import obj_array_vectorize
+import numpy as np
+
+
+def componentwise_norm(discr, a):
+    """Calculate a component-wise infinity norm."""
+    if isinstance(a, ConservedVars):
+        return componentwise_norm(discr, a.join())
+    return obj_array_vectorize(lambda b: discr.norm(b, np.inf), a)
+
+
+def componentwise_err(discr, lhs, rhs, relative=True):
+    """Calculate the component-wise error."""
+    lhs_norm = componentwise_norm(discr, lhs)
+    rhs_norm = componentwise_norm(discr, rhs)
+
+    err_norm = componentwise_norm(discr, rhs - lhs)
```

Comment:
This is not how you measure error. The error between 1 and -1 would be 0 under your formula.

Reply:
?

Reply:
I don't think it should- if lhs = [1], rhs = [-1], lhs norm would be 1, rhs norm should be 1, rhs - lhs = [-2] so its norm is 2, so relative error should be 2/1 = 2. I'll add a unit test to make sure, though.

Reply:
If I'm understanding your concern correctly, the assertion added in (010265c) should check the case you brought up. I'm subtracting arrays, not subtracting norms, so it should be able to tell apart [1] and [-1].